pkg/timelock/timelock.go

```diff
@@ -109,182 +102,269 @@ func NewTimelockWorker(nodeURL, timelockAddress, callProxyAddress, privateKey st
+	go func() {
+		defer close(done)
+		defer close(logCh)
+		defer sub.Unsubscribe()
+
+		for {
+			select {
+			case err := <-sub.Err():
+				// Check if the error is not nil, because sub.Unsubscribe will
+				// signal the channel sub.Err() to close it, leading to false nil errors.
+				if err != nil {
+					tw.logger.Warn().Msgf("subscription error: %s", err.Error())
+					SetReadyStatus(HealthStatusError)
+					sub.Unsubscribe()
+
+					success := false
+					for try := range maxSubRetries {
+						tw.logger.Warn().Msgf("trying to re-create subscription: %v/%v retry.", try+1, maxSubRetries)
+						sub, err = tw.ethClient.SubscribeFilterLogs(ctx, query, logCh)
+						if err == nil {
+							tw.logger.Info().Msg("subscription successfully recreated.")
+							SetReadyStatus(HealthStatusOK)
+							success = true
+							break
+						}
+
+						time.Sleep(time.Second * time.Duration(try))
+					}
+
+					if !success {
+						tw.logger.Error().Msg("failed to recreate subscription after retries: shutting down timelock-worker.")
+						return
+					}
+				}
+
+			case <-ctx.Done():
+				tw.logger.Debug().Msgf("shutting down subscription")
+				SetReadyStatus(HealthStatusError)
+				return
+			}
+		}
+	}()
+
+	return done, logCh, nil
+}
+
+// retrieveHistoricalLogs returns a types.Log channel and retrieves all the historical events of a given contract.
+// Once all the logs have been sent into the channel the function returns and the channel is closed.
+func (tw *Worker) retrieveHistoricalLogs(ctx context.Context) (<-chan struct{}, <-chan types.Log, error) {
+	query := tw.setupFilterQuery()
+	logCh := make(chan types.Log)
+	done := make(chan struct{})
+
+	// FilterLogs starts filtering the logs at fromBlock, gathering the historical data.
+	// This is needed to guarantee that all the events are gathered, even in scenarios where the worker crashes.
 	filter, err := tw.ethClient.FilterLogs(ctx, query)
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
 
+	// Process incoming historical logs in a separate goroutine.
 	go func() {
-		for _, l := range filter {
-			logCh <- l
+		defer close(done)
+		defer close(logCh)
+
+		for _, log := range filter {
+			select {
+			case logCh <- log:
+				tw.logger.Debug().Msgf("processing historical log: %+v\n", log)
+			case <-ctx.Done():
+				tw.logger.Debug().Msg("stopped while processing historical logs: incomplete retrieval.")
+				return
+			}
 		}
+		tw.logger.Debug().Msg("retrieved correctly all historical logs")
 	}()
 
-	// Setting readyStatus here because we want to make sure subscription is up.
-	tw.logger.Info().Msgf("Initial subscription complete")
-	SetReadyStatus(HealthStatusOK)
+	return done, logCh, nil
+}
+
+// processLogs is implemented as a fan-in for all the logs channels, merging all the data and handling logs sequentially.
+// This function is thread safe.
+func (tw *Worker) processLogs(ctx context.Context, oldLog, newLog <-chan types.Log) <-chan struct{} {
```

Comment:
There's a scenario where both log channels are closed but the context is not canceled. In that case we should stop processing. 

One way is to merge the two log channels into a single channel and exit the processing loop either when the log channel is closed or the context is cancelled. 

In this case we need to stop the Listen method either when we're done processing or the context is cancelled. 

Right now we only clean up on context cancellation.